### ISSUE TITLE
Akka stream Source.queue hangs when using fail overflow strategy

### ISSUE THREAD

**Issue by edrevo** [akka/akka#23078]
Repro code:

```scala
val queue = 
  Source.queue[Unit](10, OverflowStrategy.fail)
    .throttle(1, 1 second, 1, ThrottleMode.shaping)
    .to(Sink.ignore)
    .run()

Await.result(
  (1 to 15).map(_ => queue.offer(())).last,
  Duration.Inf)
```

Expected behaviour: the future should end with failure
Observed behavior: the Await never returns

**Comment by johanandren** [comment]
The problem is that the stream fails (since that is what you said it should do on overflow), so when subsequent elements are enqueued, they are sent to a stopped stream, which will never complete the future. The logic is here: https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/impl/Sources.scala#L180

If you are going to use Fail as overflow strategy you will need to not offer a new element until the previous future completed with failure or success.

**Comment by jrudolph** [comment]
I think it's a bug. It should be possible to keep enough state in the asynchronous boundary to fix those edge cases (also #23081).

**Comment by jrudolph** [comment]
The problem here is that stages can lose async signals that are already scheduled while the stage shuts down. I created https://github.com/akka/akka/issues/23111 to track this issue. As the bug is in `CallbackWrapper` it might make sense to manage the queue differently for `QueueSource` to avoid the problem in the first place.

**Comment by johanandren** [comment]
Jungle of related tickets here, think it is good to mention #20503 as well.

**Comment by johanandren** [comment]
This should be fixed by the fix for #23185
